lms/services/grading_info.py

```diff
@@ -49,10 +49,20 @@ def get_students_for_grading(
                 # In LTI 1.3 lis_result_sourcedid == user_id
                 # or rather the concept of lis_result_sourcedid doesn't really exists and the LTI1.3 grading API is based on the user id.
                 # We take the user id value instead here for LTI1.3.
-                # This is important in the case of upgrades that happen midterm, wih grading_infos from before the upgrade:
+                # This is important in the case of upgrades that happen midterm, with grading_infos from before the upgrade:
                 # we might have only the LTI1.1 value for lis_result_sourcedid but if we pick the user id instead
                 # we are guaranteed to get the right value for the LTI1.3 API
                 lis_result_sourced_id = grading_info.user_id
+                if application_instance.settings.get(
+                    "hypothesis", "lti_13_sourcedid_for_grading", False
+                ):
+                    # While all the above is true the situation gets further complicated by the  lti1p1 claim that provides
+                    # the old values of some IDs in LTI1.3
+                    # In this scenario, for students that have launched after the upgrade will get:
+                    #  grading_info.user_id to be the old  LTI user ID, the LTI1.1
+                    #  grading_info.lis_result_sourced_id will be the "sub" parameter via LTIParams
+                    # In that case grading_info.lis_result_sourced_id is the right value to use to talk to the LTI1.3 grading API
+                    lis_result_sourced_id = grading_info.lis_result_sourcedid
```

Comment:
The factoring here is a bit extra verbose, we assign one value, for LTI1.1, then if in 1.3 we use the other and if this setting is enabled we go back to the original one.

I reckon in this case the extra verbosity best conveys what's going on.